sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceEventClient.java

```diff
@@ -0,0 +1,402 @@
+            } else {
+                // No: We're just not actively running anything. Long decline until we have services.
+                return OfferResponse.processed(recommendations);
+            }
+        } else if (anyClientsNotReady) {
+            // One or more clients said they weren't ready. Tell upstream to short-decline the unused offers, but still
+            // perform any operations returned by the ready clients.
+            return OfferResponse.notReady(recommendations);
+        } else {
+            // We have one or more clients and they were all able to process offers, so tell upstream to long-decline.
+            return OfferResponse.processed(recommendations);
+        }
+    }
+
+    /**
+     * Maps the reserved resources in the provided unused offers according to the services that own them, then queries
+     * those services directly to see what resources they consider unexpected.
+     *
+     * <p>This is an optimization which avoids querying services about unexpected resources that don't relate to them.
+     * <p>In addition to reducing unnecessary queries, this also improves isolation between services. They only see
+     * resources which relate to them.
+     */
+    @Override
+    public UnexpectedResourcesResponse getUnexpectedResources(Collection<Protos.Offer> unusedOffers) {
+        // Resources can be unexpected for any of the following reasons:
+        // - (CASE 1) Resources which lack a service name (shouldn't happen in practice)
+        // - (CASE 2) Resources with an unrecognized service name (old resources?)
+        // - (CASE 3) Resources whose matching service returned them as unexpected (old/decommissioned resources?)
+
+        // For each offer, the resources which should be unreserved.
+        Map<Protos.OfferID, OfferResources> unexpectedResources = new HashMap<>();
+        // For each service, the resources associated with that service, paired with their parent offer(s)
+        // In other words: serviceName => offerId => offer + [resourcesForService]
+        Map<String, Map<Protos.OfferID, OfferResources>> offersByService = new HashMap<>();
+
+        // Map reserved resources (and their parent offers) to the service that they're assigned to.
+        // We can then query those services with the subset of offers that belong to them.
+        for (Protos.Offer offer : unusedOffers) {
+            for (Protos.Resource resource : offer.getResourcesList()) {
+                Optional<String> serviceName = ResourceUtils.getNamespace(resource);
+                if (serviceName.isPresent()) {
+                    // Found service name: Store resource against serviceName+offerId to be evaluated below.
+                    getEntry(offersByService, serviceName.get(), offer).add(resource);
+                } else if (ResourceUtils.getReservation(resource).isPresent()) {
+                    // (CASE 1) Malformed: Reserved resource which is missing a service name.
```

Comment:
This is safe because this code only ever runs in the multi-service case?

Reply:
Yes, `MultiServiceEventClient` is only relevant when running in multi-service mode. It's effectively slotting in where an `AbstractScheduler` which go in a mono-service situation. (Both implement `MesosEventClient`)

Reply:
Can we just ignore malformed resources?  I'm nervous about eating other frameworks resources again.

Reply:
From convo elsewhere:
- if service name label is present and maps to known service: map to service
- if service name label is present but maps to unknown service: assume it's from a valid service which was since removed
- if service name label isn't present: uh-oh, better leave this one alone